jspm_packages/npm/npm@2.14.5/man/man1/npm.1
.TH "NPM" "1" "September 2015" "" ""
.SH "NAME"
\fBnpm\fR \- javascript package manager
.SH SYNOPSIS
.P
.RS 2
.nf
npm <command> [args]
.fi
.RE
.SH VERSION
.P
2.14.5
.SH DESCRIPTION
.P
npm is the package manager for the Node JavaScript platform\.  It puts
modules in place so that node can find them, and manages dependency
conflicts intelligently\.
.P
It is extremely configurable to support a wide variety of use cases\.
Most commonly, it is used to publish, discover, install, and develop node
programs\.
.P
Run \fBnpm help\fP to get a list of available commands\.
.SH INTRODUCTION
.P
You probably got npm because you want to install stuff\.
.P
Use \fBnpm install blerg\fP to install the latest version of "blerg"\.  Check out
npm help \fBnpm\-install\fP for more info\.  It can do a lot of stuff\.
.P
Use the \fBnpm search\fP command to show everything that's available\.
Use \fBnpm ls\fP to show everything you've installed\.
.SH DEPENDENCIES
.P
If a package references to another package with a git URL, npm depends
on a preinstalled git\.
.P
If one of the packages npm tries to install is a native node module and
requires compiling of C++ Code, npm will use
node\-gyp \fIhttps://github\.com/TooTallNate/node\-gyp\fR for that task\.
For a Unix system, node\-gyp \fIhttps://github\.com/TooTallNate/node\-gyp\fR
needs Python, make and a buildchain like GCC\. On Windows,
Python and Microsoft Visual Studio C++ is needed\. Python 3 is
not supported by node\-gyp \fIhttps://github\.com/TooTallNate/node\-gyp\fR\|\.
For more information visit
the node\-gyp repository \fIhttps://github\.com/TooTallNate/node\-gyp\fR and
the node\-gyp Wiki \fIhttps://github\.com/TooTallNate/node\-gyp/wiki\fR\|\.
.SH DIRECTORIES
.P
See npm help 5 \fBnpm\-folders\fP to learn about where npm puts stuff\.
.P
In particular, npm has two modes of operation:
.RS 0
.IP \(bu 2
global mode:
.br
npm installs packages into the install prefix at
\fBprefix/lib/node_modules\fP and bins are installed in \fBprefix/bin\fP\|\.
.IP \(bu 2
local mode:
.br
npm installs packages into the current project directory, which
defaults to the current working directory\.  Packages are installed to
\fB\|\./node_modules\fP, and bins are installed to \fB\|\./node_modules/\.bin\fP\|\.

.RE
.P
Local mode is the default\.  Use \fB\-\-global\fP or \fB\-g\fP on any command to
operate in global mode instead\.
.SH DEVELOPER USAGE
.P
If you're using npm to develop and publish your code, check out the
following help topics:
.RS 0
.IP \(bu 2
json:
Make a package\.json file\.  See npm help 5 \fBpackage\.json\fP\|\.
.IP \(bu 2
link:
For linking your current working code into Node's path, so that you
don't have to reinstall every time you make a change\.  Use
\fBnpm link\fP to do this\.
.IP \(bu 2
install:
It's a good idea to install things if you don't need the symbolic link\.
Especially, installing other peoples code from the registry is done via
\fBnpm install\fP
.IP \(bu 2
adduser:
Create an account or log in\.  Credentials are stored in the
user config file\.
.IP \(bu 2
publish:
Use the \fBnpm publish\fP command to upload your code to the registry\.

.RE
.SH CONFIGURATION
.P
npm is extremely configurable\.  It reads its configuration options from
5 places\.
.RS 0
.IP \(bu 2
Command line switches:
.br
Set a config with \fB\-\-key val\fP\|\.  All keys take a value, even if they
are booleans (the config parser doesn't know what the options are at
the time of parsing\.)  If no value is provided, then the option is set
to boolean \fBtrue\fP\|\.
.IP \(bu 2
Environment Variables:
.br
Set any config by prefixing the name in an environment variable with
\fBnpm_config_\fP\|\.  For example, \fBexport npm_config_key=val\fP\|\.
.IP \(bu 2
User Configs:
.br
The file at $HOME/\.npmrc is an ini\-formatted list of configs\.  If
present, it is parsed\.  If the \fBuserconfig\fP option is set in the cli
or env, then that will be used instead\.
.IP \(bu 2
Global Configs:
.br
The file found at \.\./etc/npmrc (from the node executable, by default
this resolves to /usr/local/etc/npmrc) will be parsed if it is found\.
If the \fBglobalconfig\fP option is set in the cli, env, or user config,
then that file is parsed instead\.
.IP \(bu 2
Defaults:
.br
npm's default configuration options are defined in
lib/utils/config\-defs\.js\.  These must not be changed\.

.RE
.P
See npm help 7 \fBnpm\-config\fP for much much more information\.
.SH CONTRIBUTIONS
.P
Patches welcome!
.RS 0
.IP \(bu 2
code:
Read through npm help 7 \fBnpm\-coding\-style\fP if you plan to submit code\.
You don't have to agree with it, but you do have to follow it\.
.IP \(bu 2
docs:
If you find an error in the documentation, edit the appropriate markdown
file in the "doc" folder\.  (Don't worry about generating the man page\.)

.RE
.P
Contributors are listed in npm's \fBpackage\.json\fP file\.  You can view them
easily by doing \fBnpm view npm contributors\fP\|\.
.P
If you would like to contribute, but don't know what to work on, check
the issues list or ask on the mailing list\.
.RS 0
.IP \(bu 2
http://github\.com/npm/npm/issues
.IP \(bu 2
npm\-@googlegroups\.com

.RE
.SH BUGS
.P
When you find issues, please report them:
.RS 0
.IP \(bu 2
web:
http://github\.com/npm/npm/issues
.IP \(bu 2
email:
npm\-@googlegroups\.com

.RE
.P
Be sure to include \fIall\fR of the output from the npm command that didn't work
as expected\.  The \fBnpm\-debug\.log\fP file is also helpful to provide\.
.P
You can also look for isaacs in #node\.js on irc://irc\.freenode\.net\.  He
will no doubt tell you to put the output in a gist or email\.
.SH AUTHOR
.P
Isaac Z\. Schlueter \fIhttp://blog\.izs\.me/\fR ::
isaacs \fIhttps://github\.com/isaacs/\fR ::
@izs \fIhttp://twitter\.com/izs\fR ::
i@izs\.me
.SH SEE ALSO
.RS 0
.IP \(bu 2
npm help help
.IP \(bu 2
npm help 7 faq
.IP \(bu 2
README
.IP \(bu 2
npm help 5 package\.json
.IP \(bu 2
npm help install
.IP \(bu 2
npm help config
.IP \(bu 2
npm help 7 config
.IP \(bu 2
npm help 5 npmrc
.IP \(bu 2
npm help 7 index
.IP \(bu 2
npm apihelp npm

.RE



jspm_packages/npm/npm@2.14.5/lib/npm.js
/* */ 
(function(process) {
  ;
  (function() {
    if (typeof WScript !== "undefined") {
      WScript.echo("npm does not work when run\n" + "with the Windows Scripting Host\n\n" + "'cd' to a different directory,\n" + "or type 'npm.cmd <args>',\n" + "or type 'node npm <args>'.");
      WScript.quit(1);
      return;
    }
    var EventEmitter = require("events").EventEmitter,
        npm = module.exports = new EventEmitter(),
        npmconf = require("./config/core"),
        log = require("npmlog"),
        gfs = require("graceful-fs"),
        fs = gfs.gracefulify(require("fs")),
        path = require("path"),
        abbrev = require("abbrev"),
        which = require("which"),
        CachingRegClient = require("./cache/caching-client"),
        charSpin = require("char-spinner");
    npm.config = {
      loaded: false,
      get: function() {
        throw new Error('npm.load() required');
      },
      set: function() {
        throw new Error('npm.load() required');
      }
    };
    npm.commands = {};
    npm.rollbacks = [];
    try {
      var j = JSON.parse(fs.readFileSync(path.join(__dirname, "../package.json")) + "");
      npm.version = j.version;
    } catch (ex) {
      try {
        log.info("error reading version", ex);
      } catch (er) {}
      npm.version = ex;
    }
    var commandCache = {},
        aliases = {
          "rm": "uninstall",
          "r": "uninstall",
          "un": "uninstall",
          "unlink": "uninstall",
          "remove": "uninstall",
          "rb": "rebuild",
          "list": "ls",
          "la": "ls",
          "ll": "ls",
          "ln": "link",
          "i": "install",
          "isntall": "install",
          "up": "update",
          "upgrade": "update",
          "c": "config",
          "dist-tags": "dist-tag",
          "info": "view",
          "show": "view",
          "find": "search",
          "s": "search",
          "se": "search",
          "author": "owner",
          "home": "docs",
          "issues": "bugs",
          "unstar": "star",
          "apihelp": "help",
          "login": "adduser",
          "add-user": "adduser",
          "tst": "test",
          "t": "test",
          "find-dupes": "dedupe",
          "ddp": "dedupe",
          "v": "view",
          "verison": "version"
        },
        aliasNames = Object.keys(aliases),
        cmdList = ["install", "uninstall", "cache", "config", "set", "get", "update", "outdated", "prune", "pack", "dedupe", "rebuild", "link", "publish", "star", "stars", "tag", "adduser", "logout", "unpublish", "owner", "access", "team", "deprecate", "shrinkwrap", "help", "help-search", "ls", "search", "view", "init", "version", "edit", "explore", "docs", "repo", "bugs", "faq", "root", "prefix", "bin", "whoami", "dist-tag", "ping", "test", "stop", "start", "restart", "run-script", "completion"],
        plumbing = ["build", "unbuild", "xmas", "substack", "visnup"],
        littleGuys = ["isntall"],
        fullList = cmdList.concat(aliasNames).filter(function(c) {
          return plumbing.indexOf(c) === -1;
        }),
        abbrevs = abbrev(fullList);
    fullList = npm.fullList = fullList.filter(function(c) {
      return littleGuys.indexOf(c) === -1;
    });
    npm.spinner = {
      int: null,
      started: false,
      start: function() {
        if (npm.spinner.int)
          return;
        var c = npm.config.get("spin");
        if (!c)
          return;
        var stream = npm.config.get("logstream");
        var opt = {
          tty: c !== "always",
          stream: stream
        };
        opt.cleanup = !npm.spinner.started;
        npm.spinner.int = charSpin(opt);
        npm.spinner.started = true;
      },
      stop: function() {
        clearInterval(npm.spinner.int);
        npm.spinner.int = null;
      }
    };
    Object.keys(abbrevs).concat(plumbing).forEach(function addCommand(c) {
      Object.defineProperty(npm.commands, c, {
        get: function() {
          if (!loaded)
            throw new Error("Call npm.load(config, cb) before using this command.\n" + "See the README.md or cli.js for example usage.");
          var a = npm.deref(c);
          if (c === "la" || c === "ll") {
            npm.config.set("long", true);
          }
          npm.command = c;
          if (commandCache[a])
            return commandCache[a];
          var cmd = require(__dirname + "/" + a + ".js");
          commandCache[a] = function() {
            var args = Array.prototype.slice.call(arguments, 0);
            if (typeof args[args.length - 1] !== "function") {
              args.push(defaultCb);
            }
            if (args.length === 1)
              args.unshift([]);
            npm.registry.version = npm.version;
            if (!npm.registry.refer) {
              npm.registry.refer = [a].concat(args[0]).map(function(arg) {
                if (arg && arg.match && arg.match(/\/|\\/)) {
                  return "[REDACTED]";
                } else {
                  return arg;
                }
              }).filter(function(arg) {
                return arg && arg.match;
              }).join(" ");
            }
            cmd.apply(npm, args);
          };
          Object.keys(cmd).forEach(function(k) {
            commandCache[a][k] = cmd[k];
          });
          return commandCache[a];
        },
        enumerable: fullList.indexOf(c) !== -1,
        configurable: true
      });
      if (c.match(/\-([a-z])/)) {
        addCommand(c.replace(/\-([a-z])/g, function(a, b) {
          return b.toUpperCase();
        }));
      }
    });
    function defaultCb(er, data) {
      if (er)
        console.error(er.stack || er.message);
      else
        console.log(data);
    }
    npm.deref = function(c) {
      if (!c)
        return "";
      if (c.match(/[A-Z]/))
        c = c.replace(/([A-Z])/g, function(m) {
          return "-" + m.toLowerCase();
        });
      if (plumbing.indexOf(c) !== -1)
        return c;
      var a = abbrevs[c];
      if (aliases[a])
        a = aliases[a];
      return a;
    };
    var loaded = false,
        loading = false,
        loadErr = null,
        loadListeners = [];
    function loadCb(er) {
      loadListeners.forEach(function(cb) {
        process.nextTick(cb.bind(npm, er, npm));
      });
      loadListeners.length = 0;
    }
    npm.load = function(cli, cb_) {
      if (!cb_ && typeof cli === "function")
        cb_ = cli, cli = {};
      if (!cb_)
        cb_ = function() {};
      if (!cli)
        cli = {};
      loadListeners.push(cb_);
      if (loaded || loadErr)
        return cb(loadErr);
      if (loading)
        return;
      loading = true;
      var onload = true;
      function cb(er) {
        if (loadErr)
          return;
        loadErr = er;
        if (er)
          return cb_(er);
        if (npm.config.get("force")) {
          log.warn("using --force", "I sure hope you know what you are doing.");
        }
        npm.config.loaded = true;
        loaded = true;
        loadCb(loadErr = er);
        if (onload = onload && npm.config.get("onload-script")) {
          require(onload);
          onload = false;
        }
      }
      log.pause();
      load(npm, cli, cb);
    };
    function load(npm, cli, cb) {
      which(process.argv[0], function(er, node) {
        if (!er && node.toUpperCase() !== process.execPath.toUpperCase()) {
          log.verbose("node symlink", node);
          process.execPath = node;
          process.installPrefix = path.resolve(node, "..", "..");
        }
        var builtin = path.resolve(__dirname, "..", "npmrc");
        npmconf.load(cli, builtin, function(er, config) {
          if (er === config)
            er = null;
          npm.config = config;
          if (er)
            return cb(er);
          if (!config.get("global") && config.sources.project && config.sources.project.type !== "ini") {
            log.verbose("config", "Skipping project config: %s. " + "(matches userconfig)", config.localPrefix + "/.npmrc");
          }
          var ua = config.get("user-agent") || "";
          ua = ua.replace(/\{node-version\}/gi, process.version);
          ua = ua.replace(/\{npm-version\}/gi, npm.version);
          ua = ua.replace(/\{platform\}/gi, process.platform);
          ua = ua.replace(/\{arch\}/gi, process.arch);
          config.set("user-agent", ua);
          var color = config.get("color");
          log.level = config.get("loglevel");
          log.heading = config.get("heading") || "npm";
          log.stream = config.get("logstream");
          switch (color) {
            case "always":
              log.enableColor();
              npm.color = true;
              break;
            case false:
              log.disableColor();
              npm.color = false;
              break;
            default:
              var tty = require("tty");
              if (process.stdout.isTTY)
                npm.color = true;
              else if (!tty.isatty)
                npm.color = true;
              else if (tty.isatty(1))
                npm.color = true;
              else
                npm.color = false;
              break;
          }
          log.resume();
          npm.registry = new CachingRegClient(npm.config);
          var umask = npm.config.get("umask");
          npm.modes = {
            exec: 0777 & (~umask),
            file: 0666 & (~umask),
            umask: umask
          };
          var gp = Object.getOwnPropertyDescriptor(config, "globalPrefix");
          Object.defineProperty(npm, "globalPrefix", gp);
          var lp = Object.getOwnPropertyDescriptor(config, "localPrefix");
          Object.defineProperty(npm, "localPrefix", lp);
          return cb(null, npm);
        });
      });
    }
    Object.defineProperty(npm, "prefix", {
      get: function() {
        return npm.config.get("global") ? npm.globalPrefix : npm.localPrefix;
      },
      set: function(r) {
        var k = npm.config.get("global") ? "globalPrefix" : "localPrefix";
        return npm[k] = r;
      },
      enumerable: true
    });
    Object.defineProperty(npm, "bin", {
      get: function() {
        if (npm.config.get("global"))
          return npm.globalBin;
        return path.resolve(npm.root, ".bin");
      },
      enumerable: true
    });
    Object.defineProperty(npm, "globalBin", {get: function() {
        var b = npm.globalPrefix;
        if (process.platform !== "win32")
          b = path.resolve(b, "bin");
        return b;
      }});
    Object.defineProperty(npm, "dir", {
      get: function() {
        if (npm.config.get("global"))
          return npm.globalDir;
        return path.resolve(npm.prefix, "node_modules");
      },
      enumerable: true
    });
    Object.defineProperty(npm, "globalDir", {
      get: function() {
        return (process.platform !== "win32") ? path.resolve(npm.globalPrefix, "lib", "node_modules") : path.resolve(npm.globalPrefix, "node_modules");
      },
      enumerable: true
    });
    Object.defineProperty(npm, "root", {get: function() {
        return npm.dir;
      }});
    Object.defineProperty(npm, "cache", {
      get: function() {
        return npm.config.get("cache");
      },
      set: function(r) {
        return npm.config.set("cache", r);
      },
      enumerable: true
    });
    var tmpFolder;
    var rand = require("crypto").randomBytes(4).toString("hex");
    Object.defineProperty(npm, "tmp", {
      get: function() {
        if (!tmpFolder)
          tmpFolder = "npm-" + process.pid + "-" + rand;
        return path.resolve(npm.config.get("tmp"), tmpFolder);
      },
      enumerable: true
    });
    Object.getOwnPropertyNames(npm.commands).forEach(function(n) {
      if (npm.hasOwnProperty(n) || n === "config")
        return;
      Object.defineProperty(npm, n, {
        get: function() {
          return function() {
            var args = Array.prototype.slice.call(arguments, 0),
                cb = defaultCb;
            if (args.length === 1 && Array.isArray(args[0])) {
              args = args[0];
            }
            if (typeof args[args.length - 1] === "function") {
              cb = args.pop();
            }
            npm.commands[n](args, cb);
          };
        },
        enumerable: false,
        configurable: true
      });
    });
    if (require.main === module) {
      require("../bin/npm-cli");
    }
  })();
})(require("process"));
